### PR TITLE
keepalive: RT timeout parameters

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: ref #229

Problem Summary: When `ACK` not arrived in time, the connection will abort.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/proxy/keepalive/timeout.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+var (
+	ErrTimeout = errors.New("failed to set TCP RT timeout")
+)
+
+func SetTimeout(conn net.Conn, d time.Duration) error {
+	tcpcn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return errors.Wrapf(ErrKeepAlive, "not net.TCPConn")
```
Comment (djshow832) [inline pkg/proxy/keepalive/timeout.go:31]:
```suggestion
		return errors.Wrapf(ErrTimeout, "not net.TCPConn")
```

File: pkg/proxy/keepalive/timeout.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	syscn, err := tcpcn.SyscallConn()
+	if err != nil {
+		return errors.Wrap(ErrKeepAlive, err)
```
Comment (djshow832) [inline pkg/proxy/keepalive/timeout.go:36]:
```suggestion
		return errors.Wrap(ErrTimeout, err)
```

File: pkg/proxy/keepalive/timeout_darwin.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/pingcap/TiProxy/lib/util/errors"
+)
+
+func setTimeout(syscn syscall.RawConn, d time.Duration) error {
+	var serr error
+	return errors.Collect(ErrTimeout, serr, syscn.Control(func(fd uintptr) {
+		serr = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_RXT_CONNDROPTIME, int(d.Seconds()))
+		if serr != nil {
+			return
+		}
```
Comment (djshow832) [inline pkg/proxy/keepalive/timeout_darwin.go:32]: Why need this?
BTW, should TCP_CONNECTIONTIMEOUT be also set?

Comment (xhebox) [reply pkg/proxy/keepalive/timeout_darwin.go]: I am not sure about if `TCP_CONNECTIONTIMEOUT` is the correct parameter for apple and BSD. But `TCP_RXT_CONNDROPTIME` is more similar to `TCP_USER_TIMEOUT`.

Comment (djshow832) [reply pkg/proxy/keepalive/timeout_darwin.go]: I saw that for BSD it's TCP_KEEPINIT. https://gitlab.isc.org/isc-projects/bind9/-/issues/2334

Comment (djshow832) [reply pkg/proxy/keepalive/timeout_darwin.go]: I saw that TCP_CONNECTIONTIMEOUT is for creating new connections. We already have dial timeout, so we don't need that.

File: pkg/proxy/net/packetio_test.go [hunk 4]
```diff
@@ -268,17 +269,25 @@ func TestKeepAlive(t *testing.T) {
...
+	backendUnhealthy.Cnt = 1
+	backendUnhealthy.Intvl = time.Second
 	testTCPConn(t,
 		func(t *testing.T, cli *PacketIO) {
 			require.NoError(t, cli.SetKeepalive(frontend))
 			require.NoError(t, cli.ClientTLSHandshake(ctls))
+			// timeout by keepalive + timeout
+			time.Sleep(backendUnhealthy.Timeout)
+			_, err := cli.ReadPacket()
+			require.Error(t, err)
```
Comment (djshow832) [inline pkg/proxy/net/packetio_test.go:283]: I mean, cli sets timeout and srv sleeps 2s. And the read packet won't report error.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**